### PR TITLE
 Add Active Window Sensor Integration for Plasma 5 & 6

### PR DESCRIPTION
**Pull request by TheOddPirate**
This PR introduces a new Active Window sensor that exposes the currently focused window's title (caption or resourceClass) as a string in Home Assistant.

* **Plasma 6:** Fully tested and confirmed working on Manjaro.
* **Plasma 5:** Included a compatible script based on older documentation; untested in a live environment, added to support more systems.

The sensor registers a D-Bus service, dynamically loads a KWin script, and updates Home Assistant whenever the active window changes. This provides consistent cross-version support and extends Home Assistant's integration with KDE desktop environments.

Future improvements may include additional testing for Plasma 5 and support for fallback behavior if scripts fail to load.

* Plasma version is detected via `KDE_SESSION_VERSION`.
* Minor adjustments made for Plasma 5 vs. 6 APIs.
* The integration tries to follow the existing `Sensor` and `Integration` pattern for maintainability.

### CODE REVIEW

File: integrations/activewindow.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    QVariant arg = reply.arguments().first();
+    QString scriptObjectPath;
+
+    if (arg.canConvert<QDBusObjectPath>()) {
+        QDBusObjectPath path = arg.value<QDBusObjectPath>();
+        scriptObjectPath = path.path();
+        //qDebug() << "ActiveWindowWatcher: loadScript returned object path" << scriptObjectPath;
+    } else if (arg.canConvert<int>()) {
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: we know it returns an int

File: integrations/activewindow.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    
+    QFile scriptFile(m_scriptPath);
+    if (!scriptFile.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
+        qWarning() << "ActiveWindowWatcher: could not open script file" << m_scriptPath;
+        return false;
+    }
+
+    QTextStream out(&scriptFile);
+    QString plasmaVersion = detectPlasmaVersion();
+    if (plasmaVersion == "5") {
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: Lets kill this 5 support.  5 is long dead at this point.

File: integrations/activewindow.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+{
+    if (title != m_lastTitle) {
+        m_lastTitle = title;
+        publish(title);
+    }
+}
+
+void ActiveWindowWatcher::publish(const QString &state)
+{
+    if (!state.isEmpty()){
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: This 'if' statement doesn't make sense, especially in combination with the line below.

File: integrations/activewindow.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+void ActiveWindowWatcher::onActiveWindowChanged(const QString &title)
+{
+    if (title != m_lastTitle) {
+        m_lastTitle = title;
+        publish(title);
+    }
+}
+
+void ActiveWindowWatcher::publish(const QString &state)
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: it's not really a state, it's the title most the time.

File: integrations/activewindow.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+};
+
+ActiveWindowWatcher::ActiveWindowWatcher(QObject *parent)
+    : QObject(parent)
+{
+    m_sensor = new Sensor(this);
+    m_sensor->setId("active_window");
+    m_sensor->setName("Active Window");
+    m_sensor->setDiscoveryConfig("icon", "mdi:application");
+    m_sensor->setState("Loading...");
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: I don't think this `setState("Loading...")` adds anything useful to the HA consumer side.

File: integrations/activewindow.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    if (title != m_lastTitle) {
+        m_lastTitle = title;
+        publish(title);
+    }
+}
+
+void ActiveWindowWatcher::publish(const QString &state)
+{
+    if (!state.isEmpty()){
+    m_sensor->setState(state.isEmpty() ? "No active window" : state);
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: I would suggest having an empty string for no window is clearer.

**Review by davidedmundson — COMMENTED** [review]
Can you add

  - "--talk-name=org.kde.KWin"

to .flatpak-manifest.yaml

File: integrations/activewindow.cpp [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    Sensor *m_sensor;
+    QDBusInterface *m_kwinIface = nullptr;
+    QString m_lastTitle;
+    QString m_scriptPath;
+    bool m_connected = false;
+};
+
+// Helper for detecting Plasma version
+QString detectPlasmaVersion()
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: unused

File: integrations/activewindow.cpp [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        "        width: w.width,\n"
+        "        height: w.height,\n"
+        "        pid: w.pid\n"
+        "    };\n"
+        "\n"
+        "    var payloadStr = JSON.stringify(payload);\n"
+        "    if (payloadStr !== JSON.stringify(lastPayload)) {\n"
+        "        lastPayload = payload;\n"
+        "        callDBus('org.davidedmundson.kiot.ActiveWindow', '/ActiveWindow', 'org.davidedmundson.kiot.ActiveWindow', 'UpdateAttributes', payloadStr);\n"
+        "        callDBus('org.davidedmundson.kiot.ActiveWindow', '/ActiveWindow', 'org.davidedmundson.kiot.ActiveWindow', 'UpdateTitle', payload.title);\n"
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: why do we do both?

We have the title in the attributes

File: integrations/activewindow.cpp [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        "var currentWindow = null;\n"
+        "\n"
+        "function updateActiveWindow(w) {\n"
+        "    if (!w) return;\n"
+        "\n"
+        "    var payload = {\n"
+        "        title: w.caption || '',\n"
+        "        resourceClass: w.resourceClass || '',\n"
+        "        windowType: w.windowType || '',\n"
+        "        minimized: w.minimized,\n"
```
Comment (davidedmundson) [inline integrations/activewindow.cpp]: resourceClass makes sense.

What would you ever do on the HA side with:

 - minimized (will always be false for an active window)
 - x/y/width/height
 - pid

File: integrations/activewindow.cpp [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        qWarning() << "ActiveWindowWatcher: could not open script file" << m_scriptPath;
+        return false;
+    }
+
+    QTextStream out(&scriptFile);
+    out << "var lastPayload = {};\n"
+        "var currentWindow = null;\n"
+        "\n"
+        "function updateActiveWindow(w) {\n"
+        "    if (!w) return;\n"
```
Comment (davidedmundson) [inline integrations/activewindow.cpp:111]: something for the future.

You probably want some code to get the transient parent of w. 

Otherwise whenever we right click on something, we'll get the details of the popup not the app we care about.

If the goal was to make the lights go green when you were using Kate, that would be annoying.

File: core.cpp [hunk 11]
```diff
@@ -323,21 +323,52 @@ Sensor::Sensor(QObject *parent)
...
+{
+    if (HaControl::mqttClient()->state() != QMqttClient::Connected)
+        return;
+
+    HaControl::mqttClient()->publish(baseTopic(), m_state.toUtf8(), 0, true);
+}
+
+void Sensor::publishAttributes()
+{
+    if (m_attributes.isEmpty() || HaControl::mqttClient()->state() != QMqttClient::Connected)
```
Comment (davidedmundson) [inline core.cpp]: consider the following:

setAtributes({foo: 1});
then later.
setAtributes({});

the HA side will not reset the attributes.

File: core.cpp [hunk 12]
```diff
@@ -323,21 +323,52 @@ Sensor::Sensor(QObject *parent)
 void Sensor::init()
 {
     setHaType("sensor");
+
+    // Standard state topic
     setHaConfig({
         {"state_topic", baseTopic()},
+        {"json_attributes_topic", baseTopic() + "/attributes"} // ny topic for attributes
```
Comment (davidedmundson) [inline core.cpp:330]: did you test you were able to use these for an automation from HA?

File: integrations/activewindow.cpp (hunk 9 above)
Comment (TheOddPirate) [reply integrations/activewindow.cpp]: fullscreen was the primary attribute I needed.  
resourceClass also makes sense long term, so I kept it.  
The other attributes were included because they may be useful for some integrations on the HA side, but they are not essential for my use case.

If you prefer a more minimal API surface I can remove x/y/width/height and pid without issue. Let me know what aligns best with the intended design.

File: core.cpp (hunk 12 above)
Comment (TheOddPirate) [reply core.cpp:330]: > did you test you were able to use these for an automation from HA?

I only tested them as conditions, and there they worked, but after testing triggers now that failed, but i found the fix by making the bool .toString() in the kwin script it works. I will update with the fix asap

File: integrations/activewindow.cpp (hunk 9 above)
Comment (TheOddPirate) [reply integrations/activewindow.cpp]: So yeah, i would say resolved, but its up to you if the extra attributes should be removed or not?

File: integrations/activewindow.cpp (hunk 10 above)
Comment (TheOddPirate) [reply integrations/activewindow.cpp:111]: I’m not familiar with the term “transient parent” in the context of KWin scripting, so I’m not sure what scenario you are referring to.  
I tried reproducing it by right clicking and opening various menus in Kate, but the script still reports the main Kate window every time.

If you can point me to a concrete example where KWin exposes a transient instead of the actual application window, I can adjust the script accordingly.

Comment (TheOddPirate) [reply integrations/activewindow.cpp:111]: The main functionality is working, but I haven’t found a reliable way to test the transient parent edge case yet. I’ve added handling for the transient and transientFor properties and will refine if/when a concrete scenario appears.
